applications/serial_lte_modem/src/slm_at_tcp_proxy.c

```diff
@@ -435,123 +443,255 @@ static int tcp_data_save(uint8_t *data, uint32_t length)
 	return ring_buf_put(&data_buf, data, length);
 }
 
-static void tcp_thread_func(void *p1, void *p2, void *p3)
+static void tcp_data_handle(uint8_t *data, uint32_t length)
 {
 	int ret;
-	struct pollfd fds;
-	int sock;
 
-	ARG_UNUSED(p1);
-	ARG_UNUSED(p2);
-	ARG_UNUSED(p3);
+	if (proxy.datamode) {
+		rsp_send(data, length);
+	} else if (slm_util_hex_check(data, length)) {
+		ret = slm_util_htoa(data, length, data_hex, DATA_HEX_MAX_SIZE);
+		if (ret < 0) {
+			LOG_ERR("hex convert error: %d", ret);
+			return;
+		}
+		if (tcp_data_save(data_hex, ret) < 0) {
+			sprintf(rsp_buf, "#XTCPDATA: overrun\r\n");
+		} else {
+			sprintf(rsp_buf, "#XTCPDATA: %d, %d\r\n",
+				DATATYPE_HEXADECIMAL, ret);
+		}
+		rsp_send(rsp_buf, strlen(rsp_buf));
+	} else {
+		if (tcp_data_save(data, length) < 0) {
+			sprintf(rsp_buf, "#XTCPDATA: overrun\r\n");
+		} else {
+			sprintf(rsp_buf, "#XTCPDATA: %d, %d\r\n",
+			DATATYPE_PLAINTEXT, length);
+		}
+		rsp_send(rsp_buf, strlen(rsp_buf));
+	}
 
-thread_entry:
-	if (proxy.role == AT_TCP_ROLE_SERVER) {
-		socklen_t len = sizeof(struct sockaddr_in);
+}
+
+int tcpsvr_input(int infd)
+{
+	int ret;
+
+	if (fds[infd].fd == proxy.sock) {
+		socklen_t len;
 		char peer_addr[INET_ADDRSTRLEN];
 
-		/* Accept incoming connection */;
-		LOG_DBG("Accept connection...");
-		proxy.sock_peer = INVALID_SOCKET;
-		ret = accept(proxy.sock, (struct sockaddr *)&remote, &len);
+		len = sizeof(struct sockaddr_in);
+		ret = accept(proxy.sock,
+				(struct sockaddr *)&remote, &len);
 		if (ret < 0) {
 			LOG_ERR("accept() failed: %d", -errno);
-			do_tcp_server_stop(-errno);
-			return;
+			return -errno;
+		}
+		if (nfds >= MAX_POLL_FD) {
+			LOG_WRN("Full. Close connection.");
+			close(ret);
+			return -ECONNREFUSED;
 		}
+		/* Accept incoming connection */
+		LOG_DBG("accept(): %d", ret);
 		if (inet_ntop(AF_INET, &remote.sin_addr, peer_addr,
 			INET_ADDRSTRLEN) != NULL) {
 			sprintf(rsp_buf, "#XTCPSVR: %s connected\r\n",
 				peer_addr);
 			rsp_send(rsp_buf, strlen(rsp_buf));
 		}
 		proxy.sock_peer = ret;
+		LOG_DBG("New connection - %d",
+			proxy.sock_peer);
+		fds[nfds].fd = proxy.sock_peer;
+		fds[nfds].events = POLLIN;
+		nfds++;
 		/* Start a one-shot timer to close the connection */
-		k_timer_start(&conn_timer, K_SECONDS(CONFIG_SLM_TCP_CONN_TIME),
+		k_timer_start(&conn_timer,
+			      K_SECONDS(CONFIG_SLM_TCP_CONN_TIME),
 			      K_NO_WAIT);
-	}
-
-	if (proxy.role == AT_TCP_ROLE_SERVER) {
-		sock = proxy.sock_peer;
 	} else {
-		sock = proxy.sock;
+		char data[NET_IPV4_MTU];
+
+		k_timer_stop(&conn_timer);
+		ret = recv(fds[infd].fd, data, NET_IPV4_MTU, 0);
+		if (ret < 0) {
+			LOG_WRN("recv() error: %d", -errno);
+			return -errno;
+		}
+		if (ret == 0) {
+			goto recv_done;
+		}
+		tcp_data_handle(data, ret);
+recv_done:
+		/* Restart conn timer */
+		LOG_ERR("restart timer: POLLIN");
```

Comment:
Should this really be an error log? From what I understand it'll get printed each time the socket receives some data (I'm a bit worried it might bloat the log output even if it was at debug level).

Reply:
Yes, you are correct. It is some debug code left... I've changed LOG_ERR to LOG_DBG.